mxfusion/inference/batch_loop.py

```diff
@@ -66,3 +67,97 @@ def run(self, infr_executor, data, param_dict, ctx, optimizer='adam',
+        """
+        :param learning_rate: Not used
+        :param infr_executor: The MXNet function that computes the training objective.
+        :type infr_executor: MXNet Gluon Block
+        :param data: a list of observed variables
+        :type data: [mxnet.ndarray]
+        :param param_dict: The MXNet ParameterDict for Gradient-based optimization
+        :type param_dict: mxnet.gluon.ParameterDict
+        :param ctx: MXNet context
+        :type ctx: mxnet.cpu or mxnet.gpu
+        :param optimizer: the choice of optimizer (default: 'adam')
+        :type optimizer: str
+        :param max_iter: the maximum number of iterations of gradient optimization
+        :type max_iter: int
+        :param verbose: whether to print per-iteration messages.
+        :type verbose: boolean
+        :param logger: The logger to send logs to
+        :type logger: :class:`inference.Logger`
+        """
+
+        if self._initialized is None:
+            self._initialize(param_dict)
+
+        x0 = self._extract_array_from_param_dict(param_dict)
+        func = lambda x: self._f_df(x, infr_executor, data, ctx, param_dict)
+
+        x_opt, f_opt, _ = fmin_l_bfgs_b(func, x0=x0, maxiter=max_iter)
+        print('Final loss:', f_opt)
+
+    def _f_df(self, x, infr_executor, data, ctx, param_dict):
+        """
+        Objective function for optimization. Returns function value and gradient
+        """
+        self._set_param_values_from_array(x, param_dict)
+        with mx.autograd.record():
+            loss, loss_for_gradient = infr_executor(mx.nd.zeros(1, ctx=ctx), *data)
+            loss_for_gradient.backward()
+        return loss.asnumpy().astype('float64'), self._extract_grad(param_dict).astype('float64')
+
+    def _extract_grad(self, param_dict):
+        """
+        Extracts gradient information from parameter dictionary
+        """
+        x = mx.nd.zeros(self._i_end[-1])
+        for i, (key, val) in enumerate(param_dict.items()):
+            x[self._i_start[i]:self._i_end[i]] = val.data().grad.reshape((-1,))
+        return x.asnumpy()
+
+    def _initialize(self, param_dict):
+        """
+        Stores stuff like shapes of all the parameters so we can flatten them all to one array for scipy and reshape
+        them back to the correct shape
+        """
+        self._i_start = []
+        self._i_end = []
+        self._shapes = []
+
+        i = 0
+        for key, val in param_dict.items():
+            self._i_start.append(i)
+            self._i_end.append(i + val.data().size)
+
+            i += val.data().size
+            self._shapes.append(val.data().shape)
+
+    def _set_param_values_from_array(self, x, param_dict):
+        """
+        Sets the updated x values in the parameter dictionary
+        """
+        for i, (key, val) in enumerate(param_dict.items()):
+            value_to_set_to = mx.nd.reshape(mx.nd.array(x[self._i_start[i]:self._i_end[i]]), self._shapes[i])
```

Comment:
be careful about the dtype and ctx of the mxnet array.

Reply:
Good point. I'll extract the dtype and context of the parameters along with the shapes